8322036: Improve help output from the javadoc tool

nizarbenalla opened the pull request:
Could I please get a review for this small clenup? I updated the `--help` command output to match that of `man javadoc`.

I also noticed that  `--help-extended` was renamed to `--help-extra` at some point but doc-comment was not changed, so I changed it.
I could create a separate issue for it if needed, just to keep a record of it.

The new `javadoc --help` output is the following 

```
javadoc --help
Usage:
    javadoc [options] [packagenames] [sourcefiles] [@files]
where options include:
    @<file>       Read options and filenames from file
    --add-modules <module>(,<module>)*
                  Root modules to resolve in addition to the initial modules,
                  or all modules on the module path if <module> is
                  ALL-MODULE-PATH.
    -bootclasspath <path>
                  Override location of platform class files used for non-modular
                  releases
    -breakiterator
                  Compute first sentence with BreakIterator
    --class-path <path>, -classpath <path>, -cp <path>
                  Specify where to find user class files
    -doclet <class>
                  Generate output via alternate doclet
    -docletpath <path>
                  Specify where to find doclet class files
    --enable-preview
                  Enable preview language features. To be used in conjunction with
                  either -source or --release.
    -encoding <name>
                  Source file encoding name
    -exclude <pkglist>
                  Specify a list of packages to exclude
    --expand-requires (transitive|all)
                  Instructs the tool to expand the set of modules to be
                  documented. By default, only the modules given explicitly on
                  the command line will be documented. A value of "transitive"
                  will additionally include all "requires transitive"
                  dependencies of those modules. A value of "all" will include
                  all dependencies of those modules.
    -extdirs <dirlist>
                  Override location of installed extensions
    --help, -help, -?, -h
                  Display command-line options and exit
    --help-extra, -X
                  Print a synopsis of nonstandard options and exit
    -J<flag>      Pass <flag> directly to the runtime system
    --limit-modules <module>(,<module>)*
                  Limit the universe of observable modules
    -locale <name>
                  Locale to be used, e.g. en_US or en_US_WIN
    --module <module>(,<module>)*
                  Document the specified module(s)
    --module-path <path>, -p <path>
                  Specify where to find application modules
    --module-source-path <path>
                  Specify where to find input source files for multiple modules
    -package
                  Show package/protected/public types and members. For 
                  named modules, show all packages and all module details.
    -private
                  Show all types and members. For named modules,
                  show all packages and all module details.
    -protected
                  Show protected/public types and members (default). For
                  named modules, show exported packages and the module's API.
    -public
                  Show only public types and members. For named modules,
                  show exported packages and the module's API.
    -quiet        Do not display status messages
    --release <release>
                  Provide source compatibility with specified release
    --show-members <value>
                  Specifies which members (fields, methods, or constructors) will be
                  documented, where value can be one of "public", "protected",
                  "package" or "private". The default is "protected", which will
                  show public and protected members, "public" will show only
                  public members, "package" will show public, protected and
                  package members and "private" will show all members.
    --show-module-contents <value>
                  Specifies the documentation granularity of module
                  declarations. Possible values are "api" or "all".
    --show-packages <value>
                  Specifies which module packages will be documented. Possible
                  values are "exported" or "all" packages.
    --show-types <value>
                  Specifies which types (classes, interfaces, etc.) will be
                  documented, where value can be one of "public", "protected",
                  "package" or "private". The default is "protected", which will
                  show public and protected types, "public" will show only
                  public types, "package" will show public, protected and
                  package types and "private" will show all types.
    --source <release>, -source <release>
                  Provide source compatibility with specified release
    --source-path <path>, -sourcepath <path>
                  Specify where to find source files
    -subpackages <subpkglist>
                  Specify subpackages to recursively load
    --system <jdk>
                  Override location of system modules used for modular releases
    --upgrade-module-path <path>
                  Override location of upgradeable modules
    -verbose      Output messages about what Javadoc is doing
    --version     Print version information
    -Werror       Report an error if any warnings occur

Provided by the Standard doclet:
    --add-script <file>
                  Add a script file to the generated documentation
    --add-stylesheet <file>
                  Add a stylesheet file to the generated documentation
    --allow-script-in-comments
                  Allow JavaScript in documentation comments, and options
                  whose value is _html-code_
    -author       Include @author paragraphs
    -bottom <html-code>
                  Include bottom text for each page
    -charset <charset>
                  Charset for cross-platform viewing of generated documentation
    -d <directory>
                  Destination directory for output files
    -docencoding <name>
                  Specify the character encoding for the output
    -docfilessubdirs
                  Enables deep copying of `doc-files` directories. Subdirectories and all
                  contents are recursively copied to the destination
    -doctitle <html-code>
                  Include title for the overview page
    -excludedocfilessubdir <name>,<name>,...
                  Exclude any `doc-files` subdirectories with given name.
                  ':' can also be used anywhere in the argument as a separator.
    -footer <html-code>
                  This option is no longer supported and reports a warning
    -group <name> <g1>,<g2>...
                  Group specified elements together in overview page.
                  ':' can also be used anywhere in the argument as a separator.
    -header <html-code>
                  Include header text for each page
    -helpfile <file>
                  Specifies a file containing the text that will be displayed when the
                  help link in the navigation bar is clicked
    -html5        Generate HTML 5 output. This option is no longer required.
    --javafx, -javafx
                  Enable JavaFX functionality
    -keywords     Include HTML meta tags with package, class and member info
    -link <url>   Create links to javadoc output at <url>
    --link-modularity-mismatch (warn|info)
                  Report external documentation with wrong modularity with either
                  a warning or informational message. The default behaviour is to
                  report a warning.
    -linkoffline <url1> <url2>
                  Link to docs at <url1> using package list at <url2>
    --link-platform-properties <url>
                  Link to platform documentation URLs declared in properties file at <url>
    -linksource   Generate source in HTML
    --main-stylesheet <file>, -stylesheetfile <file>
                  File to change style of the generated documentation
    -nocomment    Suppress description and tags, generate only declarations
    -nodeprecated
                  Do not include @deprecated information
    -nodeprecatedlist
                  Do not generate deprecated list
    --no-fonts    Do not include standard web fonts in generated documentation
    -nohelp       Do not generate help link
    -noindex      Do not generate index
    -nonavbar     Do not generate navigation bar
    --no-platform-links
                  Do not generate links to the platform documentation
    -noqualifier <name1>,<name2>,...
                  Exclude the list of qualifiers from the output.
                  ':' can also be used anywhere in the argument as a separator.
    -nosince      Do not include @since information
    -notimestamp  Do not include hidden time stamp
    -notree       Do not generate class hierarchy
    --override-methods (detail|summary)
                  Document overridden methods in the detail or summary sections.
                  The default is 'detail'.
    -overview <file>
                  Read overview documentation from HTML file
    -serialwarn   Reports compile-time warnings for missing `@serial` tags
    --since <release>(,<release>)*
                  Document new and deprecated API in the specified releases
    --since-label <text>
                  Provide text to use in the heading of the "New API" page
    --snippet-path <path>
                  The path for external snippets
    -sourcetab <tab length>
                  Specify the number of spaces each tab takes up in the source
    --spec-base-url 
                  Specify a base URL for relative URLs in @spec tags
    -splitindex   Split index into one file per letter
    -tag <name>:<locations>:<header>
                  Specifies a custom tag with a single argument
    -taglet       The fully qualified name of Taglet to register
    -tagletpath   The path to Taglets
    -top <html-code>
                  Include top text for each page
    -use          Create class and package usage pages
    -version      Include @version paragraphs
    -windowtitle <text>
                  Browser window title for the documentation

GNU-style options may use = instead of whitespace to separate the name of an
option from its value.
```

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8322036](https://bugs.openjdk.org/browse/JDK-8322036): Improve help output from the javadoc tool (**Bug** - P4)


### Reviewers
 * [Pavel Rappo](https://openjdk.org/census#prappo) (@pavelrappo - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20618/head:pull/20618` \
`$ git checkout pull/20618`

Update a local copy of the PR: \
`$ git checkout pull/20618` \
`$ git pull https://git.openjdk.org/jdk.git pull/20618/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20618`

View PR using the GUI difftool: \
`$ git pr show -t 20618`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20618.diff">https://git.openjdk.org/jdk/pull/20618.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20618#issuecomment-2293686481)

pavelrappo commented:
Nizar, thanks for doing this.

> this small clenup

It's not a cleanup, it's updating help output. It's important to be price here. Cleanup is optional, updating incorrect/inaccurate help output is not.

> I also noticed that `--help-extended` was renamed to `--help-extra` at some point but doc-comment was not changed, so I changed it.

Good catch! That said, I don't think there has ever been the `--help-extended` option or alias. It seems that `--help-extra` has been the correct alias for `-X` option for tools such as java, javac, javadoc, and jshell for as long as I can see. It's just javadoc inexplicably renamed it to `--help-extended` in [8237803][]. Later, `--help-extended` also appeared in [8294546][].

My guess, is that it's how @jonathan-gibbons (mis)remembered `-X`'s alias in his head. To Jon's credit, the relevant `OptionKind.EXTENDED` enum name is confusing. But since the above PRs were reviewed by me and @jddarcy respectively, if there's anyone to blame, it's us.

If my guess is correct, those two occurrences of `--help-extended` should be deleted. One is deleted in this PR, another should be deleted by compiler people. I'll haven't yet finished looking at this PR. If I have more comments, I'll post them later. Thanks.

[8237803]: https://mail.openjdk.org/pipermail/javadoc-dev/2020-January/001324.html	"(sorry, pre git)"
[8294546]: https://github.com/openjdk/jdk/pull/10566

pavelrappo reviewed (commented):
Outside this PR diff, do we need to lowercase this ["Standard Doclet"] similarly to how it is done in JDK-8290702, whose updates this PR carries over to javadoc help output?

["Standard Doclet"]: https://github.com/openjdk/jdk/blob/d0a265039a36292d87b249af0e8977982e5acc7b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties#L26-L25

jonathan-gibbons commented:
> Outside this PR diff, do we need to lowercase this ["Standard Doclet"](https://github.com/openjdk/jdk/blob/d0a265039a36292d87b249af0e8977982e5acc7b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties#L26-L25) similarly to how it is done in JDK-8290702, whose updates this PR carries over to javadoc help output?

I still have mixed thoughts on this. 
The name of the doclet has always been "Standard Doclet": that is a proper noun and should be capitalized. The lower case form, "standard doclet" is just a description.

That being said, I agree we should be consistent.

jonathan-gibbons commented:
> My guess, is that it's how @jonathan-gibbons (mis)remembered -X's alias in his head.

mea culpa.

nizarbenalla commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/ToolOptions.java:
```diff
@@ -678,7 +678,7 @@ interface ShowHelper {
         /**
          * Show command-line help for the extended options, as requested by
-         * the {@code --help-extended} option and its aliases.
+         * the {@code --help-extra} option and its aliases.
```
Thanks!

nizarbenalla commented:
These new changes address the review comments. I replaced backticks with single quotes and dropped the unnecessary use of underscores.

I capitalized the first letter of "doclet" in `Provided by the Standard doclet:` , as "Standard" is also capitalized [here](https://github.com/openjdk/jdk/blob/6041c936d6dd39c5b3a89ed2823b25a8aef42b9f/src/jdk.javadoc/share/classes/jdk/javadoc/doclet/StandardDoclet.java#L91)

The new help output is the following 

```
javadoc --help
Usage:
    javadoc [options] [packagenames] [sourcefiles] [@files]
where options include:
    @<file>       Read options and filenames from file
    --add-modules <module>(,<module>)*
                  Root modules to resolve in addition to the initial modules,
                  or all modules on the module path if <module> is
                  ALL-MODULE-PATH.
    -bootclasspath <path>
                  Override location of platform class files used for non-modular
                  releases
    -breakiterator
                  Compute first sentence with BreakIterator
    --class-path <path>, -classpath <path>, -cp <path>
                  Specify where to find user class files
    -doclet <class>
                  Generate output via alternate doclet
    -docletpath <path>
                  Specify where to find doclet class files
    --enable-preview
                  Enable preview language features. To be used in conjunction with
                  either -source or --release.
    -encoding <name>
                  Source file encoding name
    -exclude <pkglist>
                  Specify a list of packages to exclude
    --expand-requires (transitive|all)
                  Instructs the tool to expand the set of modules to be
                  documented. By default, only the modules given explicitly on
                  the command line will be documented. A value of "transitive"
                  will additionally include all "requires transitive"
                  dependencies of those modules. A value of "all" will include
                  all dependencies of those modules.
    -extdirs <dirlist>
                  Override location of installed extensions
    --help, -help, -?, -h
                  Display command-line options and exit
    --help-extra, -X
                  Print a synopsis of nonstandard options and exit
    -J<flag>      Pass <flag> directly to the runtime system
    --limit-modules <module>(,<module>)*
                  Limit the universe of observable modules
    -locale <name>
                  Locale to be used, e.g. en_US or en_US_WIN
    --module <module>(,<module>)*
                  Document the specified module(s)
    --module-path <path>, -p <path>
                  Specify where to find application modules
    --module-source-path <path>
                  Specify where to find input source files for multiple modules
    -package
                  Show package/protected/public types and members. For 
                  named modules, show all packages and all module details.
    -private
                  Show all types and members. For named modules,
                  show all packages and all module details.
    -protected
                  Show protected/public types and members (default). For
                  named modules, show exported packages and the module's API.
    -public
                  Show only public types and members. For named modules,
                  show exported packages and the module's API.
    -quiet        Do not display status messages
    --release <release>
                  Provide source compatibility with specified release
    --show-members <value>
                  Specifies which members (fields, methods, or constructors) will be
                  documented, where value can be one of "public", "protected",
                  "package" or "private". The default is "protected", which will
                  show public and protected members, "public" will show only
                  public members, "package" will show public, protected and
                  package members and "private" will show all members.
    --show-module-contents <value>
                  Specifies the documentation granularity of module
                  declarations. Possible values are "api" or "all".
    --show-packages <value>
                  Specifies which module packages will be documented. Possible
                  values are "exported" or "all" packages.
    --show-types <value>
                  Specifies which types (classes, interfaces, etc.) will be
                  documented, where value can be one of "public", "protected",
                  "package" or "private". The default is "protected", which will
                  show public and protected types, "public" will show only
                  public types, "package" will show public, protected and
                  package types and "private" will show all types.
    --source <release>, -source <release>
                  Provide source compatibility with specified release
    --source-path <path>, -sourcepath <path>
                  Specify where to find source files
    -subpackages <subpkglist>
                  Specify subpackages to recursively load
    --system <jdk>
                  Override location of system modules used for modular releases
    --upgrade-module-path <path>
                  Override location of upgradeable modules
    -verbose      Output messages about what Javadoc is doing
    --version     Print version information
    -Werror       Report an error if any warnings occur

Provided by the Standard Doclet:
    --add-script <file>
                  Add a script file to the generated documentation
    --add-stylesheet <file>
                  Add a stylesheet file to the generated documentation
    --allow-script-in-comments
                  Allow JavaScript in documentation comments, and options
                  whose value is html-code
    -author       Include @author paragraphs
    -bottom <html-code>
                  Include bottom text for each page
    -charset <charset>
                  Charset for cross-platform viewing of generated documentation
    -d <directory>
                  Destination directory for output files
    -docencoding <name>
                  Specify the character encoding for the output
    -docfilessubdirs
                  Enables deep copying of 'doc-files' directories. Subdirectories and all
                  contents are recursively copied to the destination
    -doctitle <html-code>
                  Include title for the overview page
    -excludedocfilessubdir <name>,<name>,...
                  Exclude any 'doc-files' subdirectories with given name.
                  ':' can also be used anywhere in the argument as a separator.
    -footer <html-code>
                  This option is no longer supported and reports a warning
    -group <name> <g1>,<g2>...
                  Group specified elements together in overview page.
                  ':' can also be used anywhere in the argument as a separator.
    -header <html-code>
                  Include header text for each page
    -helpfile <file>
                  Specifies a file containing the text that will be displayed when the
                  help link in the navigation bar is clicked
    -html5        Generate HTML 5 output. This option is no longer required.
    --javafx, -javafx
                  Enable JavaFX functionality
    -keywords     Include HTML meta tags with package, class and member info
    -link <url>   Create links to javadoc output at <url>
    --link-modularity-mismatch (warn|info)
                  Report external documentation with wrong modularity with either
                  a warning or informational message. The default behaviour is to
                  report a warning.
    -linkoffline <url1> <url2>
                  Link to docs at <url1> using package list at <url2>
    --link-platform-properties <url>
                  Link to platform documentation URLs declared in properties file at <url>
    -linksource   Generate source in HTML
    --main-stylesheet <file>, -stylesheetfile <file>
                  File to change style of the generated documentation
    -nocomment    Suppress description and tags, generate only declarations
    -nodeprecated
                  Do not include @deprecated information
    -nodeprecatedlist
                  Do not generate deprecated list
    --no-fonts    Do not include standard web fonts in generated documentation
    -nohelp       Do not generate help link
    -noindex      Do not generate index
    -nonavbar     Do not generate navigation bar
    --no-platform-links
                  Do not generate links to the platform documentation
    -noqualifier <name1>,<name2>,...
                  Exclude the list of qualifiers from the output.
                  ':' can also be used anywhere in the argument as a separator.
    -nosince      Do not include @since information
    -notimestamp  Do not include hidden time stamp
    -notree       Do not generate class hierarchy
    --override-methods (detail|summary)
                  Document overridden methods in the detail or summary sections.
                  The default is 'detail'.
    -overview <file>
                  Read overview documentation from HTML file
    -serialwarn   Reports compile-time warnings for missing '@serial' tags
    --since <release>(,<release>)*
                  Document new and deprecated API in the specified releases
    --since-label <text>
                  Provide text to use in the heading of the "New API" page
    --snippet-path <path>
                  The path for external snippets
    -sourcetab <tab length>
                  Specify the number of spaces each tab takes up in the source
    --spec-base-url 
                  Specify a base URL for relative URLs in @spec tags
    -splitindex   Split index into one file per letter
    -tag <name>:<locations>:<header>
                  Specifies a custom tag with a single argument
    -taglet       The fully qualified name of Taglet to register
    -tagletpath   The path to Taglets
    -top <html-code>
                  Include top text for each page
    -use          Create class and package usage pages
    -version      Include @version paragraphs
    -windowtitle <text>
                  Browser window title for the documentation

GNU-style options may use = instead of whitespace to separate the name of an
option from its value.
```

nizarbenalla commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties:
```diff
@@ -478,7 +478,8 @@ doclet.usage.author.description=\
 
 doclet.usage.docfilessubdirs.description=\
-    Recursively copy doc-file subdirectories
+    Enables deep copying of `doc-files` directories. Subdirectories and all\n\
```
Fixed in [e24bbd8](https://github.com/openjdk/jdk/pull/20618/commits/e24bbd8d6c1f55f2e43ad453703c26105d0b4e7a), thanks.

pavelrappo replied:
@nizarbenalla have you checked that the help output is rendered correctly after you changed backticks to single quotes?

nizarbenalla replied:
Yes, I checked locally on my Mac machine and the output looks correct. ie
```
....
    -overview <file>
                  Read overview documentation from HTML file
    -serialwarn   Reports compile-time warnings for missing '@serial' tags
    --since <release>(,<release>)*
                  Document new and deprecated API in the specified releases
....
```

pavelrappo commented:
> > Outside this PR diff, do we need to lowercase this ["Standard Doclet"](https://github.com/openjdk/jdk/blob/d0a265039a36292d87b249af0e8977982e5acc7b/src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/resources/standard.properties#L26-L25) similarly to how it is done in JDK-8290702, whose updates this PR carries over to javadoc help output?
> 
> I still have mixed thoughts on this. The name of the doclet has always been "Standard Doclet": that is a proper noun and should be capitalized. The lower case form, "standard doclet" is just a description.
> 
> That being said, I agree we should be consistent.

Jon, is it possible to figure it out once and for all? It's time consuming to discuss it anew every time a new case is found. Are you saying that these two cases, which are currently present in the mainline, are okay because the meaning is slightly different?

    ...
    doclet.build_version=Standard Doclet version {0}
    ...
    doclet.help.footnote=\
    This help file applies to API documentation generated by the standard doclet.

pavelrappo commented on src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/resources/javadoc.properties:
```diff
@@ -277,7 +277,7 @@ main.opt.disable.line.doc.comments.desc=\
 These options are non-standard and subject to change without notice.
 
-main.doclet.usage.header=\nProvided by the {0} doclet:
+main.doclet.usage.header=\nProvided by the {0} Doclet:
```
I'm unsure about this one: `{0}` does not necessarily expand to `Standard`.

nizarbenalla replied:
I can change it back.
By the way I only see it used in the codebase once [here](https://github.com/openjdk/jdk/blob/20d8f58c92009a46dfb91b951e7d87b4cb8e8b41/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java#L254) where `doclet.getName()` returns `Standard`

pavelrappo replied:
> I can change it back. By the way I only see it used in the codebase once [here](https://github.com/openjdk/jdk/blob/20d8f58c92009a46dfb91b951e7d87b4cb8e8b41/src/jdk.javadoc/share/classes/jdk/javadoc/internal/tool/Start.java#L254) where `doclet.getName()` returns `Standard`

`doclet.getName()` returns a name of the doclet in use. It's not always "Standard". Check that example doclet from `jdk.javadoc/jdk.javadoc.doclet` javadoc: it returns "Example".

nizarbenalla replied:
I understand what you mean now. Thanks!

pavelrappo reviewed (approved):
Thanks for doing this. Looks good.

While reading help output, I noticed other things we could improve in a separate PR. In particular, every option should provide three elements: a synopsis of what it is for, a set of possible values, and the default value.

nizarbenalla commented:
Thank you for the reviews.

I did notice that some options had short but more informative descriptions in the man page vs the --help option.

/integrate

pavelrappo commented:
/sponsor
